### PR TITLE
Check timeout of token when validating access tokens

### PR DESCRIPTION
**Pull request by haaakon**
OAuth tests now running, but not on travis because it still has no support for Swift 1.2. The tests can be run on the "Digipost" target. 

Also added some cleanup for targets and schemes.

### CODE REVIEW

File: Digipost-Test-Dpost Tests/OAuthTests.swift [hunk 1]
```diff
@@ -17,9 +18,9 @@ class OAuthTests: XCTestCase {
     }
     
     override func tearDown() {
-        LUKeychainAccess.standardKeychainAccess().deleteAll()
-        // Put teardown code here. This method is called after the invocation of each test method in the class.
         super.tearDown()
+        OAuthToken.removeAllTokens()
+        // Put teardown code here. This method is called after the invocation of each test method in the class.
```
Comment (lillesand) [inline Digipost-Test-Dpost Tests/OAuthTests.swift]: Highly removable comment ;)

File: Digipost-Test-Dpost Tests/OAuthTests.swift [hunk 2]
```diff
@@ -44,34 +45,41 @@ class OAuthTests: XCTestCase {
         let token = OAuthToken(attributes: oAuthDictionary, scope: scope)
         return token!
     }
-    
+
     func testOauthFromDictionary() {
```
Comment (lillesand) [inline Digipost-Test-Dpost Tests/OAuthTests.swift]: Could we use more descriptive test names? What are we really testing here?

File: Digipost-Test-Dpost Tests/OAuthTests.swift [hunk 3]
```diff
@@ -44,34 +45,41 @@ class OAuthTests: XCTestCase {
...
-        let anotherToken = OAuthToken(attributes: invalidAuthDictionary, scope: "anotherScope")
-        XCTAssertNil(anotherToken, "token should not have been created")
     }
     
     func testOauthFromStrings() {
+        LUKeychainAccess.standardKeychainAccess().errorHandler = self
         let accesstoken = "accesstoken"
-        let token = OAuthToken(refreshToken: "refresh", accessToken:accesstoken, scope: kOauth2ScopeFull)
-        
+        let fierje = OAuthToken.oAuthTokenWithScope(kOauth2ScopeFull)
```
Comment (lillesand) [inline Digipost-Test-Dpost Tests/OAuthTests.swift]: What is a fierje? :S

File: Digipost/APIClient+Document.swift [hunk 4]
```diff
@@ -21,28 +19,20 @@ extension APIClient {
...
-        let task = urlSessionTask(httpMethod.post, url: document.updateUri, parameters: parameters, success: success) { (error) -> () in
-            if (error.code == Constants.Error.Code.oAuthUnathorized ) {
-                self.changeName(document, newName: name, success: success, failure: failure)
-            } else {
-                failure(error: error)
-            }
+
+        validateFullScope {
+            let task = self.urlSessionTask(httpMethod.post, url: document.updateUri, parameters: parameters, success: success,failure: failure)
+            task?.resume()
```
Comment (lillesand) [inline Digipost/APIClient+Document.swift]: `task` should probably never be Nil anyways, so we could just force unwrap and let it :fire: if it for some incomprehensible reason is. Then we at least will know about it, rather than failing silently.

File: Digipost/APIClient+Document.swift [hunk 5]
```diff
@@ -21,28 +19,20 @@ extension APIClient {
...
     func deleteDocument(uri: String, success: () -> Void , failure: (error: APIError) -> ()) {
-        let task = urlSessionTask(httpMethod.delete, url: uri, success: success) { (error) -> Void in
-            if error.code == Constants.Error.Code.oAuthUnathorized {
-                self.deleteDocument(uri, success: success, failure: failure)
-            } else {
-                failure(error: error)
-            }
+        validateFullScope {
+            let task = self.urlSessionTask(httpMethod.delete, url: uri, success: success, failure: failure)
+            task?.resume()
```
Comment (lillesand) [inline Digipost/APIClient+Document.swift]: Force unwrap

File: Digipost/APIClient+Document.swift [hunk 6]
```diff
@@ -52,37 +42,24 @@ extension APIClient {
...
             }()
-        let task = urlSessionTask(httpMethod.post, url: document.updateUri, parameters:parameters, success: success) { (error) -> () in
-            if error.code == Constants.Error.Code.oAuthUnathorized.rawValue {
-                self.moveDocument(document, toFolder: folder, success: success, failure: failure)
-            } else {
-                failure(error: error)
-            }
+        validateFullScope {
+            let task = self.urlSessionTask(httpMethod.post, url: document.updateUri, parameters:parameters, success: success, failure: failure)
+            task?.resume()
```
Comment (lillesand) [inline Digipost/APIClient+Document.swift]: Force unwrap

… and so on for the rest of them

Comment (haaakon) [inline Digipost/APIClient+Document.swift]: agreed, force app to crash if no task is created

File: Digipost-Test-Dpost Tests/OAuthTests.swift (hunk 3 above)
Comment (haaakon) [inline Digipost-Test-Dpost Tests/OAuthTests.swift]: probably autocomplete gone wrong

File: Digipost-Test-Dpost Tests/OAuthTests.swift (hunk 2 above)
Comment (haaakon) [inline Digipost-Test-Dpost Tests/OAuthTests.swift]: agreed, i'll add more specific names. For example testCreateOAuthTokenFromJsonDictionary

File: Digipost-Test-Dpost Tests/OAuthTests.swift (hunk 1 above)
Comment (haaakon) [inline Digipost-Test-Dpost Tests/OAuthTests.swift]: Yes

File: Digipost/APIClient.swift [hunk 7]
```diff
@@ -99,17 +99,18 @@ class APIClient : NSObject, NSURLSessionTaskDelegate, NSURLSessionDelegate, NSUR
         return urlString
     }
 
-    func validateTokensThenPerformTask(task: NSURLSessionTask) {
-        validateOAuthToken(kOauth2ScopeFull) {
-            self.updateAuthorizationHeader(kOauth2ScopeFull)
+    func validate(#token: OAuthToken?, thenPerformTask task: NSURLSessionTask) {
```
Comment (lillesand) [inline Digipost/APIClient.swift]: Is this used from outside this class? Is it used at all..? Could it be private?

Comment (haaakon) [inline Digipost/APIClient.swift]: I think i can remove the method, its basically deprecated and replaced with the more clear validate:then { }

File: Digipost/POSLetterViewController.m [hunk 8]
```diff
@@ -167,31 +166,41 @@ - (void)viewDidLoad
 - (void)shouldValidateOpeningReceipt:(POSAttachment *)attachment
 {
     POSDocument *document = attachment.document;
-    [[APIClient sharedClient] validateOpeningReceipt:attachment success:^{
```
Comment (haaakon) [inline Digipost/POSLetterViewController.m:170]: All changes after this seems to be triggered by an update in Clang-format http://clang.llvm.org/docs/ClangFormat.html
It's used to keep the objc code nicely formatted (though some times changes alot of the code if a new rule is added)
Optimally code changes like that should be committed as own commits to not hide their actual purpose

File: Digipost/APIClient.swift (hunk 7 above)
Comment (lillesand) [inline Digipost/APIClient.swift]: Excellent! Remove at will :+1:

**Comment by lillesand** [comment]
Generic question: what happens if a request fails now? E.g. if we loose sync with the server

File: Digipost/APIClient.swift [hunk 9]
```diff
@@ -124,72 +125,47 @@ class APIClient : NSObject, NSURLSessionTaskDelegate, NSURLSessionDelegate, NSUR
...
-                failure(error: error)
-            }
+        validateFullScope {
+            let task = self.urlSessionTask(httpMethod.post, url: invoice.sendToBankUri, success: success, failure: failure)
+            task?.resume()
         }
-        validateTokensThenPerformTask(task!)
     }
 
     func updateReceiptsInMailboxWithDigipostAddress(digipostAddress: String, uri: String, success: (Dictionary<String,AnyObject>) -> Void , failure: (error: APIError) -> ()) {
```
Comment (lillesand) [inline Digipost/APIClient.swift:150]: Isn't this really a generic `postJSON` function now? It does the same as `updateBankAccount`. Not sure if it matters as the code probably is more readable this way. Any reason why we accept the URL as an argument here when we don't other places..?

**Comment by haaakon** [comment]
User gets an error message.

File: Digipost/APIClient.swift [hunk 10]
```diff
@@ -352,6 +330,7 @@ class APIClient : NSObject, NSURLSessionTaskDelegate, NSURLSessionDelegate, NSUR
             }, error: nil)
         urlRequest.setValue("*/*", forHTTPHeaderField: "Accept")
         fileTransferSessionManager.setTaskDidCompleteBlock { (session, task, error) -> Void in
+
```
Comment (lillesand) [inline Digipost/APIClient.swift:321]: Does this do anything?

File: Digipost/APIClient.swift [hunk 11]
```diff
@@ -360,26 +339,26 @@ class APIClient : NSObject, NSURLSessionTaskDelegate, NSURLSessionDelegate, NSUR
                 self.uploadProgress?.completedUnitCount = totalSent
             })
         }
+        validateFullScope { () -> Void in
+            let task = self.fileTransferSessionManager.dataTaskWithRequest(urlRequest, completionHandler: { (response, anyObject, error) -> Void in
+                dispatch_async(dispatch_get_main_queue(), {
+                    self.removeTemporaryUploadFiles()
+                    self.isUploadingFile = false
+                    if self.isUnauthorized(response as! NSHTTPURLResponse?) {
```
Comment (lillesand) [inline Digipost/APIClient.swift:335]: Is this necessary when we validate scope first?

Comment (haaakon) [inline Digipost/APIClient.swift:335]: Probably not, the error will be handled by the other error hander. I can remove it!

File: Digipost/APIClient.swift (hunk 10 above)
Comment (haaakon) [inline Digipost/APIClient.swift:321]: No, Alamofire calls both the closures. This will be fixed when removing alamofire

File: Digipost/APIClient.swift [hunk 12]
```diff
@@ -433,9 +381,15 @@ class APIClient : NSObject, NSURLSessionTaskDelegate, NSURLSessionDelegate, NSUR
                         self.deleteRefreshTokensAndLogoutUser()
                     }
             })
-        }else if (oAuthToken == nil) {
+        } else if (oAuthToken == nil) {
             assert(false," something wrong with oauth token")
         } else {
```
Comment (lillesand) [inline Digipost/APIClient.swift:372]: I have a bit of a hard time figuring out the logic that brings us here… if we have an OAuthToken without a refresh token, it means that we're trying to jump to a higher level oauth token? Why's that? Could we make the code more explicit?

File: Digipost/APIClient.swift [hunk 13]
```diff
@@ -433,9 +381,15 @@ class APIClient : NSObject, NSURLSessionTaskDelegate, NSURLSessionDelegate, NSUR
                         self.deleteRefreshTokensAndLogoutUser()
                     }
             })
-        }else if (oAuthToken == nil) {
+        } else if (oAuthToken == nil) {
             assert(false," something wrong with oauth token")
```
Comment (lillesand) [inline Digipost/APIClient.swift]: If this is something that will never happen we should fail harder.

File: Digipost/APIClient.swift [hunk 14]
```diff
@@ -447,6 +401,18 @@ class APIClient : NSObject, NSURLSessionTaskDelegate, NSURLSessionDelegate, NSUR
...
+            letterViewController.attachment = nil
+            letterViewController.receipt = nil
+        }
+        APIClient.sharedClient.logout()
+        OAuthToken.removeAllTokens()
+        POSModelManager.sharedManager().deleteAllObjects()
+        NSNotificationCenter.defaultCenter().postNotificationName(kShowLoginViewControllerNotificationName, object: nil)
+    }
+
     func cancelUpdatingReceipts() {
```
Comment (lillesand) [inline Digipost/APIClient.swift]: Can this function be deleted?

File: Digipost/APIClient.swift (hunk 12 above)
Comment (haaakon) [inline Digipost/APIClient.swift:372]: For example if using scope idporten4 and refreshing list. It will retry with a lower level scope, "the current highest scope valid" which means 3, fullhighAuth or full, if any of those are present.

Or if you ask to read a letter with fullhighauth, the token is expired, and you retry request with a scope idporten4.

File: Digipost/APIClient.swift (hunk 13 above)
Comment (haaakon) [inline Digipost/APIClient.swift]: Yes, basically means you're logged out

File: Digipost/APIClient.swift (hunk 14 above)
Comment (haaakon) [inline Digipost/APIClient.swift]: No it deletes refresh tokens that are revoked and logs out the user. This should still be an action the app performs

File: Digipost/OAuthToken.swift [hunk 15]
```diff
@@ -37,10 +37,85 @@ class OAuthToken: NSObject, NSCoding, DebugPrintable, Printable{
...
+            self.refreshToken = acutalRefreshToken
+        } else {
+            return nil
+        }
+
+        self.scope = scope
+        storeInKeyChain()
+    }
+
+    convenience init?(refreshToken: String?, accessToken: String?, scope: String, expiresInSeconds: NSNumber) {
```
Comment (lillesand) [inline Digipost/OAuthToken.swift:74]: Could these initializers call each others? Right now we duplicate quite a lot of functionality between them.

Comment (haaakon) [inline Digipost/OAuthToken.swift:74]: Yeah probably

File: Digipost/APIClient.swift (hunk 14 above)
Comment (lillesand) [inline Digipost/APIClient.swift]: It does..? But it's empty. How does that work?

Comment (haaakon) [inline Digipost/APIClient.swift]: Ha ha. It's probably just that its no selection color on what line you actually commented. I thought you were talking about deleteRefreshTokensAndLogoutUser().
Yes cancelupdatingreceipts can be removed

File: Digipost/APIClient+Document.swift (hunk 4 above)
Comment (haaakon) [inline Digipost/APIClient+Document.swift]: Removed it being optional, might be something from an older swift version, because i can't really see the reason them being optional at all

File: Digipost/APIClient.swift (hunk 7 above)
Comment (haaakon) [inline Digipost/APIClient.swift]: commi:t 73ea5b19ef9803646680fe3e45c82f6fbe86e2d2

File: Digipost/APIClient+Document.swift (hunk 4 above)
Comment (haaakon) [inline Digipost/APIClient+Document.swift]: Commit: 4341a4c911402d2ee14d20330bd8ef79f118b086

File: Digipost-Test-Dpost Tests/OAuthTests.swift (hunk 1 above)
Comment (haaakon) [inline Digipost-Test-Dpost Tests/OAuthTests.swift]: commit: 1ef0627d97866e413a679a13a8569dde988320d2

File: Digipost-Test-Dpost Tests/OAuthTests.swift (hunk 2 above)
Comment (haaakon) [inline Digipost-Test-Dpost Tests/OAuthTests.swift]: commit: 092ab48f1b59f77516a15b46362d17afc063cf13

File: Digipost/OAuthToken.swift (hunk 15 above)
Comment (haaakon) [inline Digipost/OAuthToken.swift:74]: commit: 1e51ee199c7105e64b275e6f113d8c10dcfa3d02

File: Digipost/APIClient.swift (hunk 11 above)
Comment (haaakon) [inline Digipost/APIClient.swift:335]: commit: d09355c5537deaa377419fb7b54b4652c6c5a765 
Added a really short timeout so its easy to test refresh of access tokens
and removing test values for timeout in this commit after: 3ee2798fd9ac3721cd776ea138424c5163122319

File: Digipost/APIClient.swift (hunk 13 above)
Comment (haaakon) [inline Digipost/APIClient.swift]: I've added more comments to explain whats going on. I'm postponing the actual fail harder until i've added the error handler with logging to our own service. 
 commit: 225d5c6559219a2a2d776673eca5d5bb8f96d4dd

File: Digipost/APIClient.swift (hunk 14 above)
Comment (haaakon) [inline Digipost/APIClient.swift]: Removed them now
commit: 3e40fa48d9492b96dcf294a51dec2bb0e390fbed

File: Digipost/APIClient.swift (hunk 9 above)
Comment (haaakon) [inline Digipost/APIClient.swift:150]: Interesting, the teoretical difference is that one actually pays the invoice, other is updating the status. But seems to might be missing some implementation. Have to double check with some users that have invoices

File: Digipost/APIClient.swift (hunk 11 above)
Comment (lillesand) [inline Digipost/APIClient.swift:335]: :+1:

File: Digipost-Test-Dpost Tests/OAuthTests.swift (hunk 2 above)
Comment (lillesand) [inline Digipost-Test-Dpost Tests/OAuthTests.swift]: Much better! Very good.

**Comment by lillesand** [comment]
Good improvements, and a really good idea to reply with the commit IDs. Made it really simple to verify :+1:

**Comment by haaakon** [comment]
Glad you found it useful! Didn't take long time to fetch 'em and if it can save you time its great.
